File: go/service/institutionalholdings.go
// This file ports institutional_holdings.py: mapping a SECForm4
// /get_institution_holders payload onto Financial Datasets
// InstitutionalHolding records for get_institutional_holdings.
package service

import (
	"encoding/json"
	"sort"
	"time"

	"github.com/belazy/monid-finance/fd"
	"github.com/belazy/monid-finance/providers"
)

// institutionalHoldingsRowKeys mirrors institutional_holdings._ROW_KEYS.
var institutionalHoldingsRowKeys = []string{"results", "holders", "rows", "data", "table", "institutionHolders"}

// normalizeInstitutionalHoldings mirrors
// institutional_holdings.normalize_institutional_holdings +
// fd.institutional_holding_record, sorted by value_usd descending (rows
// without a value_usd sort last) and truncated to limit.
func normalizeInstitutionalHoldings(raw json.RawMessage, ticker string, limit int, reportPeriod dateFilters) ([]fd.InstitutionalHolding, error) {
	var value any
	if err := json.Unmarshal(raw, &value); err != nil {
		return nil, &providers.SchemaDriftError{Msg: "SECForm4 institution holders payload must be an object"}
	}
	root, ok := value.(map[string]any)
	if !ok {
		return nil, &providers.SchemaDriftError{Msg: "SECForm4 institution holders payload must be an object"}
	}
	if status, _ := root["status"].(string); status != "success" {
		return nil, &providers.SchemaDriftError{Msg: "SECForm4 institution holders status must be 'success'"}
	}
	dataValue := root["data"]
	var rows []map[string]any
	var issuerName *string
	switch data := dataValue.(type) {
	case []any:
		list, err := instHoldingsObjectList(data, "SECForm4 institution holders rows")
		if err != nil {
			return nil, err
		}
		rows = list
	case map[string]any:
		issuerName = firstStringGeneric(data, "name_of_issuer", "companyName", "company", "issuer")
		list, err := instHoldingsRows(data)
		if err != nil {
			return nil, err
		}
		rows = list
	default:
		return nil, &providers.SchemaDriftError{Msg: "SECForm4 institution holders omitted a data object"}
	}

	type scored struct {
		record   fd.InstitutionalHolding
		hasValue bool
		valueUSD int64
	}
	var records []scored
	for _, row := range rows {
		reportDay := instHoldingsFirstDate(row, "report_period", "reportDate", "date", "as_of", "periodEnd")
		if !instHoldingsMatches(reportDay, reportPeriod) {
			continue
		}
		filerName := firstStringGeneric(row, "filer_name", "institution", "holder", "manager", "name")
		shares := instHoldingsFirstInt(row, "shares", "shares_held", "num_shares", "position_shares", "current_shares")
		valueUSD := instHoldingsFirstInt(row, "value_usd", "market_value", "position_value", "current_value", "value")
		rowIssuer := firstStringGeneric(row, "name_of_issuer", "company", "issuer")
		if filerName == nil && shares == nil && valueUSD == nil {
			continue
		}
		nameOfIssuer := rowIssuer
		if nameOfIssuer == nil {
			nameOfIssuer = issuerName
		}
		tickerCopy := ticker
		record := fd.InstitutionalHolding{Ticker: &tickerCopy}
		record.NameOfIssuer = nameOfIssuer
		if reportDay != nil {
			s := reportDay.Format(dateLayout)
			record.ReportPeriod = &s
		}
		record.Shares = shares
		record.ValueUSD = valueUSD
		record.FilerName = filerName
		s := scored{record: record}
		if valueUSD != nil {
			s.hasValue = true
			s.valueUSD = *valueUSD
		}
		records = append(records, s)
	}
	// Sort by (has value_usd, value_usd) descending, mirroring the Python
	// sort key `(isinstance(value_usd, int), value_usd or 0)` under
	// reverse=True (a stable sort, so equal keys keep source order).
	sort.SliceStable(records, func(i, j int) bool {
		if records[i].hasValue != records[j].hasValue {
			return records[i].hasValue
		}
		if !records[i].hasValue {
			return false
		}
		return records[i].valueUSD > records[j].valueUSD
	})
	if limit >= 0 && len(records) > limit {
		records = records[:limit]
	}
	out := make([]fd.InstitutionalHolding, 0, len(records))
	for _, r := range records {
		out = append(out, r.record)
	}
	return out, nil
}

func instHoldingsRows(data map[string]any) ([]map[string]any, error) {
	for _, key := range institutionalHoldingsRowKeys {
		if child, ok := data[key].([]any); ok {
			return instHoldingsObjectList(child, "SECForm4 holder "+key)
		}
	}
	return nil, &providers.SchemaDriftError{Msg: "SECForm4 institution holders omitted the holder table"}
}

func instHoldingsObjectList(values []any, name string) ([]map[string]any, error) {
	rows := make([]map[string]any, 0, len(values))
	for _, item := range values {
		obj, ok := item.(map[string]any)
		if !ok {
			return nil, &providers.SchemaDriftError{Msg: name + " row must be an object"}
		}
		rows = append(rows, obj)
	}
	return rows, nil
}

func instHoldingsMatches(day *time.Time, filters dateFilters) bool {
	if !filters.any() {
		return true
	}
	if day == nil {
		return false
	}
	return filters.matches(*day)
}

func instHoldingsFirstDate(row map[string]any, keys ...string) *time.Time {
	for _, key := range keys {
		s, ok := row[key].(string)
		if !ok || s == "" {
			continue
		}
		text := s
		if len(text) > 10 {
			text = text[:10]
		}
		if t, err := time.Parse(dateLayout, text); err == nil {
			return &t
		}
	}
	return nil
}

func instHoldingsFirstInt(row map[string]any, keys ...string) *int64 {
	for _, key := range keys {
		value, ok := row[key]
		if !ok {
			continue
		}
		if _, isBool := value.(bool); isBool {
			continue
		}
		if f, ok := value.(float64); ok && f == float64(int64(f)) {
			v := int64(f)
			return &v
		}
	}
	return nil
}

File: src/monid_finance_mcp/providers/us/institutional_holdings.py
"""Financial Datasets institutional holdings via SECForm4 institution holders.

Ticker mode reads the validated SECForm4 /get_institution_holders route
and maps the reported holder rows onto Financial Datasets
InstitutionalHolding fields that the payload sources (name_of_issuer,
shares, value_usd, report_period, filer_name). Filer-CIK mode is not
routed, so the service answers with an honest bad_request.
"""
from __future__ import annotations

from datetime import date

from monid_finance_mcp.fd import institutional_holding_record
from monid_finance_mcp.models import JsonObject, JsonValue
from monid_finance_mcp.providers.us.normalize import SchemaDriftError

INSTITUTIONAL_ENDPOINT = "/get_institution_holders"

_ROW_KEYS = ("results", "holders", "rows", "data", "table", "institutionHolders")


def normalize_institutional_holdings(
    value: JsonValue,
    *,
    ticker: str,
    limit: int,
    report_period: dict[str, date | None],
) -> list[JsonObject]:
    """Map SECForm4 holder rows to FD InstitutionalHolding records, filtered."""
    root = _object(value, "SECForm4 institution holders payload")
    if root.get("status") != "success":
        raise SchemaDriftError("SECForm4 institution holders status must be 'success'")
    data = root.get("data")
    if isinstance(data, list):
        rows = _object_list(data, "SECForm4 institution holders rows")
        issuer_name: str | None = None
    elif isinstance(data, dict):
        issuer_name = _first_string(
            data, ("name_of_issuer", "companyName", "company", "issuer")
        )
        rows = _holder_rows(data)
    else:
        raise SchemaDriftError("SECForm4 institution holders omitted a data object")

    records: list[JsonObject] = []
    for index, row in enumerate(rows):
        validated = _object(row, f"SECForm4 holder row[{index}]")
        report_day = _first_date(
            validated, ("report_period", "reportDate", "date", "as_of", "periodEnd")
        )
        if not _matches(report_day, report_period):
            continue
        filer_name = _first_string(
            validated,
            ("filer_name", "institution", "holder", "manager", "name"),
        )
        shares = _first_int(
            validated,
            ("shares", "shares_held", "num_shares", "position_shares", "current_shares"),
        )
        value_usd = _first_int(
            validated,
            ("value_usd", "market_value", "position_value", "current_value", "value"),
        )
        row_issuer = _first_string(
            validated, ("name_of_issuer", "company", "issuer")
        )
        record = institutional_holding_record(
            ticker=ticker,
            name_of_issuer=row_issuer if row_issuer is not None else issuer_name,
            shares=shares,
            value_usd=value_usd,
            report_period=report_day.isoformat() if report_day is not None else None,
            filer_name=filer_name,
        )
        if filer_name is None and shares is None and value_usd is None:
            continue
        records.append(record)
    records.sort(
        key=lambda record: (
            isinstance(record.get("value_usd"), int),
            record.get("value_usd") if isinstance(record.get("value_usd"), int) else 0,
        ),
        reverse=True,
    )
    return records[:limit]


def _holder_rows(data: JsonObject) -> list[JsonObject]:
    for key in _ROW_KEYS:
        child = data.get(key)
        if isinstance(child, list):
            return _object_list(child, f"SECForm4 holder {key}")
    raise SchemaDriftError("SECForm4 institution holders omitted the holder table")


def _matches(day: date | None, report_period: dict[str, date | None]) -> bool:
    exact = report_period.get("exact")
    gte = report_period.get("gte")
    lte = report_period.get("lte")
    gt = report_period.get("gt")
    lt = report_period.get("lt")
    if exact is None and gte is None and lte is None and gt is None and lt is None:
        return True
    if day is None:
        return False
    if exact is not None and day != exact:
        return False
    if gte is not None and day < gte:
        return False
    if lte is not None and day > lte:
        return False
    if gt is not None and day <= gt:
        return False
    return lt is None or day < lt


def _object(value: JsonValue | None, name: str) -> JsonObject:
    if not isinstance(value, dict):
        raise SchemaDriftError(f"{name} must be an object")
    return value


def _object_list(value: list[JsonValue], name: str) -> list[JsonObject]:
    rows: list[JsonObject] = []
    for index, item in enumerate(value):
        if not isinstance(item, dict):
            raise SchemaDriftError(f"{name}[{index}] must be an object")
        rows.append(item)
    return rows


def _first_string(row: JsonObject, keys: tuple[str, ...]) -> str | None:
    for key in keys:
        value = row.get(key)
        if isinstance(value, str) and value:
            return value
    return None


def _first_date(row: JsonObject, keys: tuple[str, ...]) -> date | None:
    for key in keys:
        value = row.get(key)
        if not isinstance(value, str) or not value:
            continue
        try:
            return date.fromisoformat(value[:10])
        except ValueError:
            continue
    return None


def _first_int(row: JsonObject, keys: tuple[str, ...]) -> int | None:
    for key in keys:
        value = row.get(key)
        if isinstance(value, bool):
            continue
        if isinstance(value, int):
            return value
        if isinstance(value, float) and value.is_integer():
            return int(value)
    return None
